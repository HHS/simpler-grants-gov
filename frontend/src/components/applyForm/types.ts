import {
  EnumOptionsType,
  FormContextType,
  GenericObjectType,
  RJSFSchema,
  StrictRJSFSchema,
  UIOptionsType,
} from "@rjsf/utils";

import { HTMLAttributes } from "react";

export interface FormData {
  [key: string]: FormDataEntryValue;
}

export interface SetFormDataFunction {
  (data: FormData): void;
}

export interface UiSchemaField {
  type: "field";
  definition: `/properties/${string}`;
}

export interface UiSchemaSection {
  type: "section";
  label: string;
  name: string;
  number?: string;
  children: (UiSchemaField | UiSchemaSection)[];
}

export type UiSchema =
  | (UiSchemaField | UiSchemaSection)
  | (UiSchemaField | UiSchemaSection)[];

export type SchemaField = {
  type: string;
  title: string;
  minLength?: number;
  maxLength?: number;
  format?: string;
};

export interface FormSchema {
  $schema: string;
  type: string;
  properties: Record<string, SchemaField>;
  required?: string[];
}

export type TextTypes =
  | "text"
  | "email"
  | "number"
  | "password"
  | "search"
  | "tel"
  | "url";

// extends the WidgetProps type from rjsf for USWDS and this project implementation
export interface UswdsWidgetProps<
  T = unknown,
  S extends StrictRJSFSchema = RJSFSchema,
  F extends FormContextType = never,
> extends GenericObjectType,
    Pick<
      HTMLAttributes<HTMLElement>,
      Exclude<keyof HTMLAttributes<HTMLElement>, "onBlur" | "onFocus">
    > {
  id: string;
  value?: string | unknown;
  type?: string;
  minLength?: number;
  schema: RJSFSchema;
  maxLength?: number;
  required?: boolean;
  disabled?: boolean;
  readonly?: boolean;
  hideError?: boolean;
  autofocus?: boolean;
  placeholder?: string;
  options: NonNullable<UIOptionsType<T, S, F>> & {
    /** The enum options list for a type that supports them */
    enumOptions?: EnumOptionsType<S>[];
    enumDisabled?: unknown;
    emptyValue?: string;
  };
  hideLabel?: boolean;
  multiple?: boolean;
  rawErrors?: string[] | undefined;
  // whether or not to use value + onChange
  updateOnInput?: boolean;
  onChange?: (value: unknown) => void;
  onBlur?: (id: string, value: unknown) => void;
  onFocus?: (id: string, value: unknown) => void;
}
