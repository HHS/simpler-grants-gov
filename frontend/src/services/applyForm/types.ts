export interface FormData {
  [key: string]: FormDataEntryValue;
}

export interface SetFormDataFunction {
  (data: FormData): void;
}

export interface UiSchemaField {
  type: "field";
  definition: string;
}

export interface UiSchemaSection {
  type: "section";
  label: string;
  name: string;
  number: string;
  children: (UiSchemaField | UiSchemaSection)[];
}

export type UiSchema =
  | (UiSchemaField | UiSchemaSection)
  | (UiSchemaField | UiSchemaSection)[];

export type SchemaField = {
  type: string;
  title: string;
  minLength?: number;
  maxLength?: number;
  format?: string;
};

export interface FormSchema {
  title: string;
  description?: string;
  properties: [SchemaField];
  required: [string];
}

export type TextTypes =
  | "text"
  | "email"
  | "number"
  | "password"
  | "search"
  | "tel"
  | "url";
