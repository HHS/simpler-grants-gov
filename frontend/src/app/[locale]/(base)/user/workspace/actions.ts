"use server";

import { getSession } from "src/services/auth/session";
import { inviteUserToOrganization } from "src/services/fetch/fetchers/organizationsFetcher";
import { OrganizationInviteRecord } from "src/types/organizationTypes";
import { z } from "zod";

import { getTranslations } from "next-intl/server";

export type OrganizationInviteValidationErrors = {
  email?: string[];
  role?: string[];
};

export type OrganizationInviteResponse = {
  data?: OrganizationInviteRecord;
  errorMessage?: string;
  validationErrors?: OrganizationInviteValidationErrors;
  success?: boolean;
};

const validateInviteUserAction = async (formData: FormData) => {
  const t = await getTranslations("ManageUsers.inviteUser.validationErrors");
  const schema = z.object({
    email: z.string().min(1, { message: t("email") }),
    role: z.string().min(1, { message: t("role") }),
  });

  const validatedFields = schema.safeParse({
    email: formData.get("email"),
    role: formData.get("role"),
  });

  // Return early if the form data is invalid (server side validation!)
  if (!validatedFields.success) {
    return validatedFields.error.flatten().fieldErrors;
  }
};

export const inviteUserAction = async (
  _prevState: unknown,
  formData: FormData,
  organizationId: string,
): Promise<OrganizationInviteResponse> => {
  const session = await getSession();

  if (!session || !session.token || !session.user_id) {
    return {
      errorMessage: "Not logged in",
    };
  }

  const validationErrors = await validateInviteUserAction(formData);
  if (validationErrors) {
    return {
      validationErrors,
    };
  }

  const requestData = {
    email: formData.get("email") as string,
    roleId: formData.get("role") as string,
    organizationId,
  };

  let inviteUserResponse: OrganizationInviteRecord;
  try {
    inviteUserResponse = await inviteUserToOrganization(
      session.token,
      requestData,
    );
    return { data: inviteUserResponse, success: true };
  } catch (e) {
    const error = e as Error;
    console.error(
      `Error inviting user to org - ${error.message} ${error.cause?.toString() || ""}`,
    );
    return {
      errorMessage: error.message,
    };
  }
};
