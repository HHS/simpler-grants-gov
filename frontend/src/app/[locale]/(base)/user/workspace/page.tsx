import { Metadata } from "next";
import { getSession } from "src/services/auth/session";
import withFeatureFlag from "src/services/featureFlags/withFeatureFlag";
import { getUserOrganizations } from "src/services/fetch/fetchers/organizationsFetcher";
import { getUserDetails } from "src/services/fetch/fetchers/userFetcher";
import { Organization } from "src/types/applicationResponseTypes";
import { LocalizedPageProps } from "src/types/intl";

import { useTranslations } from "next-intl";
import { getTranslations } from "next-intl/server";
import Link from "next/link";
import { redirect } from "next/navigation";
import { ErrorMessage, GridContainer } from "@trussworks/react-uswds";

export async function generateMetadata({
  params,
}: LocalizedPageProps): Promise<Metadata> {
  const { locale } = await params;
  const t = await getTranslations({ locale });
  const meta: Metadata = {
    title: t("UserWorkspace.pageTitle"),
    description: t("Index.metaDescription"),
  };
  return meta;
}

const UserOrganizationsList = ({
  userOrganizations,
}: {
  userOrganizations: Organization[];
}) => {
  const t = useTranslations("UserWorkspace");
  return (
    <>
      <h2>{t("organizations")}</h2>
      <ul>
        {userOrganizations.map((userOrganization) => (
          <li key={userOrganization.organization_id}>
            <Link href={`/organization/${userOrganization.organization_id}`}>
              {userOrganization.sam_gov_entity.legal_business_name}
            </Link>
          </li>
        ))}
      </ul>
    </>
  );
};

async function UserWorkspace() {
  const t = await getTranslations("UserWorkspace");

  const session = await getSession();
  if (!session?.email) {
    // this won't happen, as email is required on sessions, and we're wrapping this in an auth gate in the layout
    console.error("no user session, or user has no email address");
    return;
  }
  let userDetails;
  let userOrganizations;
  const userDetailsPromise = getUserDetails(session.token, session.user_id);
  const userOrganizationsPromise = getUserOrganizations(
    session.token,
    session.user_id,
  );
  try {
    [userDetails, userOrganizations] = await Promise.all([
      userDetailsPromise,
      userOrganizationsPromise,
    ]);
  } catch (e) {
    console.error("Unable to fetch user details", e);
  }

  return (
    <GridContainer className="padding-top-2 tablet:padding-y-6">
      <h1>{t("title")}</h1>
      {userDetails && userOrganizations ? (
        <UserOrganizationsList userOrganizations={userOrganizations} />
      ) : (
        <ErrorMessage>{t("fetchError")}</ErrorMessage>
      )}
    </GridContainer>
  );
}

export default withFeatureFlag<object, never>(
  UserWorkspace,
  "userAdminOff",
  () => redirect("/maintenance"),
);
