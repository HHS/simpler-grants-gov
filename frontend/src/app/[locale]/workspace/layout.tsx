import { LayoutProps } from "src/types/generalTypes";

import BetaAlert from "src/components/BetaAlert";
import { AuthenticationGate } from "src/components/user/AuthenticationGate";

export default function WorkspaceLayout({ children }: LayoutProps) {
  return (
    <AuthenticationGate>
      <BetaAlert containerClasses="margin-bottom-5" />
      {children}
    </AuthenticationGate>
  );
}
