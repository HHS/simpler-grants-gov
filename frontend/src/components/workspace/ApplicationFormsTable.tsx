import { ApplicationFormDetail } from "src/types/applicationResponseTypes";
import { CompetitionForms } from "src/types/competitionsResponseTypes";
import { FormDetail } from "src/types/formResponseTypes";

import { useTranslations } from "next-intl";
import Link from "next/link";
import { Table } from "@trussworks/react-uswds";

import { USWDSIcon } from "src/components/USWDSIcon";

export const selectFormsByRequired = ({
  forms,
  required,
}: {
  forms: CompetitionForms;
  required: boolean;
}): FormDetail[] => {
  return forms.reduce<FormDetail[]>((acc, item) => {
    if (item.is_required === required) {
      acc.push(item.form);
    }
    return acc;
  }, []);
};

const selectApplicationForm = ({
  applicationForms,
  formId,
}: {
  applicationForms: ApplicationFormDetail[];
  formId: string;
}) => {
  return applicationForms
    ? applicationForms.find(
        (applicationForm) => applicationForm.form_id === formId,
      )
    : null;
};

export const ApplicationFormsTable = ({
  applicationForms,
  applicationId,
  forms,
}: {
  applicationForms: ApplicationFormDetail[];
  applicationId: string;
  forms: CompetitionForms;
}) => {
  const requiredForms = selectFormsByRequired({ forms, required: true });
  const conditionalRequiredForms = selectFormsByRequired({
    forms,
    required: false,
  });
  const t = useTranslations("Application.competitionFormTable");

  return (
    <>
      <h3>{t("requiredForms")}</h3>
      <ApplicationTable
        forms={requiredForms}
        applicationForms={applicationForms}
        applicationId={applicationId}
      />
      {conditionalRequiredForms.length > 0 && (
        <>
          <h3>{t("conditionalForms")}</h3>
          <p>{t("conditionalFormsDescription")}</p>
          <ApplicationTable
            forms={conditionalRequiredForms}
            applicationForms={applicationForms}
            applicationId={applicationId}
          />
        </>
      )}
    </>
  );
};

const selectApplicationFormById = ({
  forms,
  formId,
}: {
  forms: FormDetail[];
  formId: string;
}): FormDetail | undefined => {
  return forms.find((form) => form.form_id === formId);
};

const ApplicationTable = ({
  applicationForms,
  applicationId,
  forms,
}: {
  applicationForms: ApplicationFormDetail[];
  applicationId: string;
  forms: FormDetail[];
}) => {
  const t = useTranslations("Application.competitionFormTable");

  return (
    <Table className="width-full overflow-wrap">
      <thead>
        <tr>
          <th scope="col" className="bg-base-lightest padding-y-205">
            {t("status")}
          </th>
          <th scope="col" className="bg-base-lightest padding-y-205">
            {t("form")}
          </th>
          <th scope="col" className="bg-base-lightest padding-y-205">
            {t("instructions")}
          </th>
          <th scope="col" className="bg-base-lightest padding-y-205">
            {t("updated")}
          </th>
        </tr>
      </thead>
      <tbody>
        {applicationForms.map((form, index) => (
          <tr key={index}>
            <td data-label={t("status")}>
              <CompetitionStatus
                applicationForms={applicationForms}
                formId={form.form_id}
              />
            </td>
            <td data-label={t("form")}>
              <FormLink
                formId={form.form_id}
                forms={forms}
                applicationId={applicationId}
                appFormId={form.application_form_id}
              />
            </td>
            <td data-label={t("instructions")}>
              <InstructionsLink
                forms={forms}
                formId={form.form_id}
                text={t("downloadInstructions")}
                unavailableText={t("attachmentUnavailable")}
              />
            </td>
            <td data-label={t("updated")}>
              <div> -- </div>
            </td>
          </tr>
        ))}
      </tbody>
    </Table>
  );
};

const CompetitionStatus = ({
  formId,
  applicationForms,
}: {
  applicationForms: ApplicationFormDetail[];
  formId: string;
}) => {
  const t = useTranslations("Application.competitionFormTable.statuses");
  const applicationForm = selectApplicationForm({ formId, applicationForms });

  if (applicationForm?.application_form_status === "in_progress") {
    return (
      <div className="display-flex flex-align-center text-bold icon-active">
        <USWDSIcon name="loop" className="margin-right-2px" />
        {t("in_progress")}
      </div>
    );
  } else if (applicationForm?.application_form_status === "complete") {
    return (
      <div className="display-flex flex-align-center text-bold">
        <USWDSIcon name="check" className="text-primary margin-right-2px" />
        {t("complete")}
      </div>
    );
  } else {
    return <>-</>;
  }
};

const InstructionsLink = ({
  formId,
  forms,
  text,
  unavailableText,
}: {
  formId: string;
  forms: FormDetail[];
  text: string;
  unavailableText: string;
}) => {
  const instructions = selectApplicationFormById({
    forms,
    formId,
  })?.form_instruction;
  const downloadPath = instructions?.download_path || null;
  return (
    <>
      {downloadPath ? (
        <Link
          className="display-flex flex-align-center font-sans-2xs"
          href={downloadPath}
        >
          <USWDSIcon name="save_alt" className="margin-right-05" />
          {text}
        </Link>
      ) : (
        <>{unavailableText}</>
      )}
    </>
  );
};

const FormLink = ({
  formId,
  forms,
  applicationId,
  appFormId,
}: {
  formId: string;
  forms: FormDetail[];
  applicationId: string;
  appFormId: string;
}) => {
  const formName = selectApplicationFormById({
    forms,
    formId,
  })?.form_name;

  return (
    <>
      {formName && (
        <Link
          className="text-bold"
          href={`/workspace/applications/application/${applicationId}/form/${appFormId}`}
        >
          {formName}
        </Link>
      )}
    </>
  );
};
