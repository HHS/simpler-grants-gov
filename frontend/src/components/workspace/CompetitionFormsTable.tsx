import { CompetitionForms } from "src/types/competitionsResponseTypes";
import { FormDetail, FormInstruction } from "src/types/formResponseTypes";

import { useTranslations } from "next-intl";
import Link from "next/link";
import { Table } from "@trussworks/react-uswds";

import { USWDSIcon } from "src/components/USWDSIcon";

export const CompetitionFormsTable = ({
  forms,
  applicationId,
}: {
  forms: CompetitionForms;
  applicationId: string;
}) => {
  const requiredForms = selectForms({ forms, required: true });
  const conditionalRequiredForms = selectForms({ forms, required: false });
  const t = useTranslations("Application.competitionFormTable");

  return (
    <>
      <h3>{t("requiredForms")}</h3>

      <CompetitionTable forms={requiredForms} applicationId={applicationId} />
      <h3>{t("conditionalForms")}</h3>

      <CompetitionTable
        forms={conditionalRequiredForms}
        applicationId={applicationId}
      />
    </>
  );
};

export const selectForms = ({
  forms,
  required,
}: {
  forms: CompetitionForms;
  required: boolean;
}): FormDetail[] => {
  return forms.reduce<FormDetail[]>((acc, item) => {
    if (item.is_required === required) {
      acc.push(item.form);
    }
    return acc;
  }, []);
};

const CompetitionTable = ({
  forms,
  applicationId,
}: {
  forms: FormDetail[];
  applicationId: string;
}) => {
  const t = useTranslations("Application.competitionFormTable");

  return (
    <Table className="width-full overflow-wrap">
      <thead>
        <tr>
          <th scope="col" className="bg-base-lighter">
            {t("status")}
          </th>
          <th scope="col" className="bg-base-lighter">
            {t("form")}
          </th>
          <th scope="col" className="bg-base-lighter">
            {t("attachment")}
          </th>
          <th scope="col" className="bg-base-lighter">
            {t("instructions")}
          </th>
          <th scope="col" className="bg-base-lighter">
            {t("updated")}
          </th>
        </tr>
      </thead>
      <CompetitionTableBody forms={forms} applicationId={applicationId} />
    </Table>
  );
};

const CompetitionTableBody = ({
  forms,
  applicationId,
}: {
  forms: FormDetail[];
  applicationId: string;
}) => {
  const t = useTranslations("Application.competitionFormTable");

  return (
    <tbody>
      {forms.map((form, index) => (
        <tr key={index}>
          <td data-label={t("status")}>--</td>
          <td data-label={t("form")}>
            <Link
              className="text-bold"
              href={`/workspace/applications/application/${applicationId}/form/${form.form_id}`}
            >
              {form.form_name}
            </Link>
          </td>
          <td data-label={t("attachment")}>
            <div> -- </div>
          </td>
          <td data-label={t("instructions")}>
            <InstructionsLink
              instsruction={form.form_instruction}
              text={t("downloadInstructions")}
            />
          </td>
          <td data-label={t("updated")}>
            <div> -- </div>
          </td>
        </tr>
      ))}
    </tbody>
  );
};

const InstructionsLink = ({
  instsruction,
  text,
}: {
  instsruction: FormInstruction;
  text: string;
}) => {
  return (
    <Link
      className="display-flex flex-align-center font-sans-2xs"
      href={instsruction.download_path}
    >
      <USWDSIcon name="save_alt" className="margin-right-05" />
      {text}
    </Link>
  );
};
