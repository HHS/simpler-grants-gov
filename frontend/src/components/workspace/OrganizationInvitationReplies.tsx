import { completeStatuses, OrganizationInvitation } from "src/types/userTypes";

import { ReactNode } from "react";

import { OrganizationInvitationReply } from "./OrganizationInvitationReply";

// display prompts for all active organization invitations for a user
export const OrganizationInvitationReplies = ({
  userInvitations,
}: {
  userInvitations: OrganizationInvitation[];
}) => {
  return (
    <ul className="usa-list--unstyled">
      {userInvitations.reduce((invitationsToShow, userInvitation) => {
        if (completeStatuses.indexOf(userInvitation.status) === -1) {
          invitationsToShow.push(
            <OrganizationInvitationReply
              key={userInvitation.organization_invitation_id}
              userInvitation={userInvitation}
            />,
          );
        }
        return invitationsToShow;
      }, [] as ReactNode[])}
    </ul>
  );
};
