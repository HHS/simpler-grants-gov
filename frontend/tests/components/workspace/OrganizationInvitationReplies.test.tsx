import { render, screen } from "@testing-library/react";
import { axe } from "jest-axe";
import { completeStatuses, OrganizationInvitation } from "src/types/userTypes";
import { fakeOrganizationInvitation } from "src/utils/testing/fixtures";

import { OrganizationInvitationReplies } from "src/components/workspace/OrganizationInvitationReplies";

// mock the child so tests focus on the parent logic
jest.mock("src/components/workspace/OrganizationInvitationReply", () => ({
  OrganizationInvitationReply: ({
    userInvitation,
  }: {
    userInvitation: any;
  }) => (
    <li data-testid={`invite-${userInvitation.organization_invitation_id}`}>
      {userInvitation.organization_invitation_id} - {userInvitation.status}
    </li>
  ),
}));

const makeInvitation = (overrides: Partial<OrganizationInvitation>) => ({
  ...fakeOrganizationInvitation,
  ...overrides,
});

describe("OrganizationInvitationReplies", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("has no basic accessibility violations", async () => {
    const invites = [
      fakeOrganizationInvitation,
      makeInvitation({
        organization_invitation_id: "inv-2",
        status: "awaiting_response",
      }),
    ];
    const { container } = render(
      <OrganizationInvitationReplies userInvitations={invites} />,
    );
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });

  it("matches snapshot with multiple invitations", () => {
    const invites = [
      fakeOrganizationInvitation,
      makeInvitation({
        organization_invitation_id: "inv-2",
        status: "awaiting_response",
      }),
    ];
    const { container } = render(
      <OrganizationInvitationReplies userInvitations={invites} />,
    );
    expect(container).toMatchSnapshot();
  });

  it("does not render invitations whose status is in completeStatuses", () => {
    const completeStatus = completeStatuses[0];
    const invites = [
      fakeOrganizationInvitation,
      makeInvitation({
        organization_invitation_id: "inv-complete",
        status: completeStatus,
      }),
    ];

    render(<OrganizationInvitationReplies userInvitations={invites} />);

    // only the active invitation should be rendered
    const items = screen.getAllByRole("listitem");
    expect(items).toHaveLength(1);
    expect(screen.getByTestId("invite-uuid")).toBeInTheDocument();
    expect(screen.queryByTestId("invite-inv-complete")).not.toBeInTheDocument();
  });
});
