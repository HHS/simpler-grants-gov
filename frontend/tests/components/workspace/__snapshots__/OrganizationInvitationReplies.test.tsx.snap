// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrganizationInvitationReplies matches snapshot with multiple invitations 1`] = `
<div>
  <ul
    class="usa-list--unstyled"
  >
    <li
      data-testid="invite-uuid"
    >
      uuid
       - 
      pending
    </li>
    <li
      data-testid="invite-inv-2"
    >
      inv-2
       - 
      awaiting_response
    </li>
  </ul>
</div>
`;
